### PR TITLE
[GEOS-10129] Add language function for multilingual support in sld

### PR DESCRIPTION
**Pull request by taba90**
[GEOS-10129](https://osgeo-org.atlassian.net/browse/GEOS-10129)

Jira ticket https://osgeo-org.atlassian.net/browse/GEOS-10129?searchSessionId=ca775455-2f64-41e9-9e49-aa9311717ef3&searchObjectId=53944&searchContainerId=10000&searchContentType=issue

blocked by https://github.com/geoserver/geoserver/pull/5122

This pr add a lang function, meant to provide support to display internationalized style labels.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.**

For all pull requests:

- [ ] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md) 
- [ ] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [ ] Make sure the first PR targets the `main` branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [ ] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [ ] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

### CODE REVIEW

**Comment by aaime** [comment]
Compile failures appear to be legit:

```
Error: 0,190 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project gs-wms: Compilation failure: Compilation failure: 
Error: 0,191 [ERROR] /home/runner/work/geoserver/geoserver/src/wms/src/main/java/org/geoserver/sld/LanguageFunction.java:[8,26] cannot find symbol
Error: 0,191 [ERROR]   symbol:   class GeoServerDefaultLocale
Error: 0,192 [ERROR]   location: package org.geoserver.util
Error: 0,192 [ERROR] /home/runner/work/geoserver/geoserver/src/wms/src/main/java/org/geoserver/sld/LanguageFunction.java:[43,29] cannot find symbol
Error: 0,193 [ERROR]   symbol:   variable GeoServerDefaultLocale
Error: 0,202 [ERROR]   location: class org.geoserver.sld.LanguageFunction
```

Ah, this PR needs to be build on top of https://github.com/geoserver/geoserver/pull/5122/files#diff-802519cf7c36fb9d8d9ffc7c640a0fe521479b3f63bf7c152af43e17ca3942ae
however it does not have the necessary commits in its history.

File: src/wms/src/main/java/org/geoserver/sld/LanguageFunction.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    @Override
+    public Object evaluate(Object object) {
+        Request request = Dispatcher.REQUEST.get();
+        String result = null;
+        if (request != null
+                && request.getRawKvp() != null
+                && request.getRawKvp().containsKey(LANGUAGE)) {
+            String acceptLanguages = request.getRawKvp().get(LANGUAGE).toString();
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/wms/src/main/java/org/geoserver/sld/LanguageFunction.java]: fixed

File: src/wms/src/main/java/org/geoserver/sld/LanguageFunction.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            String[] langAr = acceptLanguages.split(" ");
+            if (langAr.length == 1) {
+                langAr = acceptLanguages.split(",");
+            }
+            result = langAr[0];
+        }
+        if (result != null && result.equals("*"))
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/wms/src/main/java/org/geoserver/sld/LanguageFunction.java]: fixed

File: src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/GetMapIntegrationTest.java [hunk 3]
```diff
@@ -2367,4 +2370,44 @@ public void testLakesWithRenderingSelection() throws Exception {
...
+                        + "&Language=en";
+        BufferedImage image = getAsImage(url, "image/png");
+        URL urlPng = getClass().getResource("multilang_result.png");
+        ImageAssert.assertEquals(new File(urlPng.toURI()), image, 500);
+    }
+
+    @Test
+    public void testMultiLanguageStyleDefaultLanguage() throws Exception {
+        WMSInfo info = getGeoServer().getService(WMSInfo.class);
+        info.setDefaultLocale(Locale.FRENCH);
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/GetMapIntegrationTest.java:2396]: fixed

File: doc/en/user/source/styling/sld/language.rst [hunk 4]
```diff
@@ -26,3 +26,29 @@ Each ``Localized`` section specifies the language (using a two letter abbreviati
 Currently, GeoServer supports localized text in SLD in WMS GetLegendGraphic requests (legends that contain labels are rendered using the
 requested language, if a ``LANGUAGE`` parameter is added to the request, e.g. ``LANGUAGE=it``).
 
+Labels in different languages
+-----------------------------
+
+GeoServer provides a ``language`` function that can be used to define ``TextSymbolizer`` with label in multiple languages. The function needs to be integrated with the ``Recode`` Transformation Function e.g:
```
[the remark this answers is not in the document]
Comment (taba90) [reply doc/en/user/source/styling/sld/language.rst]: fixed

File: doc/en/user/source/styling/sld/language.rst [hunk 5]
```diff
@@ -26,3 +26,29 @@ Each ``Localized`` section specifies the language (using a two letter abbreviati
...
+                     <ogc:Literal>name_fr</ogc:Literal>
+                   </ogc:Function>
+                 </Label>
+                 <Fill>
+                   <CssParameter name="fill">#000000</CssParameter>
+                 </Fill>
+         </TextSymbolizer>
+
+
+With such style, when peforming a GetMap adding the language parameter in the request, e.g. ``LANGUAGE=it``, if the language has been defined in the  Recode Function, the TextSymbolizer label text will be localized accordingly.
```
[the remark this answers is not in the document]
Comment (taba90) [reply doc/en/user/source/styling/sld/language.rst]: fixed

**Pull request merged**